In [413]:
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, f1_score, precision_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import LinearSVC, SVC
import warnings
warnings.filterwarnings("ignore")

In [414]:
df = pd.read_csv("Animals-FINAL-Data.csv")

In [415]:
df

Animal Name  Class             Order           Family  \
0                  Aardvark     13     Tubulidentata  Orycteropodidae   
1                  Aardwolf     13         Carnivora        Hyaenidae   
2     Abyssinian Guinea Pig     13          Rodentia         Caviidae   
3        Acadian Flycatcher      4     Passeriformes       Tyrannidae   
4                     Addax     13      Artiodactyla          Bovidae   
...                     ...    ...               ...              ...   
1436            Zebra Shark      8  Orectolobiformes  Stegostomatidae   
1437            Zebra Snake     16          Squamata         Elapidae   
1438   Zebra Spitting Cobra     16          Squamata         Elapidae   
1439                   Zebu     13      Artiodactyla          Bovidae   
1440                  Zokor     13          Rodentia       Spalacidae   

      Is_Endangered   Length(cm)  Lifespan(years)   Weight(kg)  Diet  Asia  \
0               0.0   162.500000        23.000000    70.000000     4     0   
1               0.0    93.980000        15.000000    10.909091     0     0   
2               0.0  1114.242496        10.000000  1601.274827     4     0   
3               0.0    14.478000         3.000000     0.012757     3     0   
4               1.0   124.460000        22.000000  1601.274827     2     0   
...             ...          ...              ...          ...   ...   ...   
1436            0.0  1114.242496        17.500000    18.000000     4     0   
1437            0.0   121.920000        12.000000  1601.274827     0     0   
1438            0.0  1114.242496        19.474842  1601.274827     0     0   
1439            0.0  1114.242496        14.000000   175.000000     2     1   
1440            0.0  1114.242496        19.474842     0.354369     4     1   

      Europe  Africa  South-America  North-America  Central-America  \
0          0       1              0              0                0   
1          0       1              0              0                0   
2          0       0              1              0                0   
3          0       0              1              1                1   
4          0       1              0              0                0   
...      ...     ...            ...            ...              ...   
1436       0       0              0              0                0   
1437       0       1              0              0                0   
1438       0       1              0              0                0   
1439       0       1              0              0                0   
1440       0       0              0              0                0   

      Antarctica  Oceania  Ocean  Eurasia  
0              0        0      0        0  
1              0        0      0        0  
2              0        0      0        0  
3              0        0      0        0  
4              0        0      0        0  
...          ...      ...    ...      ...  
1436           0        0      1        0  
1437           0        0      0        0  
1438           0        0      0        0  
1439           0        0      0        0  
1440           0        0      0        1  

[1441 rows x 19 columns]

In [416]:
df_edited = df.copy()

I dropped all the columns that are not numeric

In [417]:
df_edited = df.drop(labels=['Family', 'Order', 'Animal Name'], axis=1)

In [418]:
df_edited

Class  Is_Endangered   Length(cm)  Lifespan(years)   Weight(kg)  Diet  \
0        13            0.0   162.500000        23.000000    70.000000     4   
1        13            0.0    93.980000        15.000000    10.909091     0   
2        13            0.0  1114.242496        10.000000  1601.274827     4   
3         4            0.0    14.478000         3.000000     0.012757     3   
4        13            1.0   124.460000        22.000000  1601.274827     2   
...     ...            ...          ...              ...          ...   ...   
1436      8            0.0  1114.242496        17.500000    18.000000     4   
1437     16            0.0   121.920000        12.000000  1601.274827     0   
1438     16            0.0  1114.242496        19.474842  1601.274827     0   
1439     13            0.0  1114.242496        14.000000   175.000000     2   
1440     13            0.0  1114.242496        19.474842     0.354369     4   

      Asia  Europe  Africa  South-America  North-America  Central-America  \
0        0       0       1              0              0                0   
1        0       0       1              0              0                0   
2        0       0       0              1              0                0   
3        0       0       0              1              1                1   
4        0       0       1              0              0                0   
...    ...     ...     ...            ...            ...              ...   
1436     0       0       0              0              0                0   
1437     0       0       1              0              0                0   
1438     0       0       1              0              0                0   
1439     1       0       1              0              0                0   
1440     1       0       0              0              0                0   

      Antarctica  Oceania  Ocean  Eurasia  
0              0        0      0        0  
1              0        0      0        0  
2              0        0      0        0  
3              0        0      0        0  
4              0        0      0        0  
...          ...      ...    ...      ...  
1436           0        0      1        0  
1437           0        0      0        0  
1438           0        0      0        0  
1439           0        0      0        0  
1440           0        0      0        1  

[1441 rows x 16 columns]

In [419]:
df_edited.corr()

Class  Is_Endangered  Length(cm)  Lifespan(years)  \
Class            1.000000       0.137019    0.061224         0.061915   
Is_Endangered    0.137019       1.000000    0.067416         0.182451   
Length(cm)       0.061224       0.067416    1.000000         0.009590   
Lifespan(years)  0.061915       0.182451    0.009590         1.000000   
Weight(kg)       0.015696      -0.026598    0.018767        -0.023773   
Diet            -0.050547      -0.061051   -0.024429        -0.087234   
Asia             0.034269       0.044773    0.039926        -0.045219   
Europe          -0.055090      -0.094180    0.005583        -0.102752   
Africa           0.079898      -0.076095   -0.014442        -0.014504   
South-America   -0.070018      -0.065582   -0.004365        -0.073600   
North-America   -0.059874      -0.170186   -0.087244        -0.143784   
Central-America -0.101775      -0.198911   -0.033483        -0.098664   
Antarctica      -0.012681       0.023526   -0.023811         0.014134   
Oceania         -0.013325      -0.061777   -0.001159        -0.063539   
Ocean           -0.207020      -0.025998    0.027006         0.093909   
Eurasia         -0.042360      -0.090779    0.046259        -0.068345   

                 Weight(kg)      Diet      Asia    Europe    Africa  \
Class              0.015696 -0.050547  0.034269 -0.055090  0.079898   
Is_Endangered     -0.026598 -0.061051  0.044773 -0.094180 -0.076095   
Length(cm)         0.018767 -0.024429  0.039926  0.005583 -0.014442   
Lifespan(years)   -0.023773 -0.087234 -0.045219 -0.102752 -0.014504   
Weight(kg)         1.000000 -0.012555  0.014337  0.030085  0.014725   
Diet              -0.012555  1.000000  0.053867  0.079311  0.031161   
Asia               0.014337  0.053867  1.000000  0.465677  0.297003   
Europe             0.030085  0.079311  0.465677  1.000000  0.339656   
Africa             0.014725  0.031161  0.297003  0.339656  1.000000   
South-America      0.044169  0.078402  0.111577  0.215549  0.187339   
North-America      0.030934  0.071294  0.016108  0.285566 -0.002185   
Central-America    0.041773  0.090890  0.036908  0.165453  0.106958   
Antarctica        -0.011555 -0.080825  0.095803  0.138865  0.116594   
Oceania            0.024321  0.048450  0.284586  0.298191  0.260324   
Ocean              0.024622 -0.098277  0.053631  0.124917  0.046933   
Eurasia            0.023529  0.041042  0.518914  0.632189  0.326595   

                 South-America  North-America  Central-America  Antarctica  \
Class                -0.070018      -0.059874        -0.101775   -0.012681   
Is_Endangered        -0.065582      -0.170186        -0.198911    0.023526   
Length(cm)           -0.004365      -0.087244        -0.033483   -0.023811   
Lifespan(years)      -0.073600      -0.143784        -0.098664    0.014134   
Weight(kg)            0.044169       0.030934         0.041773   -0.011555   
Diet                  0.078402       0.071294         0.090890   -0.080825   
Asia                  0.111577       0.016108         0.036908    0.095803   
Europe                0.215549       0.285566         0.165453    0.138865   
Africa                0.187339      -0.002185         0.106958    0.116594   
South-America         1.000000       0.284315         0.561004    0.130427   
North-America         0.284315       1.000000         0.512298    0.050428   
Central-America       0.561004       0.512298         1.000000    0.107823   
Antarctica            0.130427       0.050428         0.107823    1.000000   
Oceania               0.271473       0.104579         0.201635    0.174226   
Ocean                 0.084600      -0.102438         0.025926    0.196520   
Eurasia               0.241733       0.252512         0.219663    0.170546   

                  Oceania     Ocean   Eurasia  
Class           -0.013325 -0.207020 -0.042360  
Is_Endangered   -0.061777 -0.025998 -0.090779  
Length(cm)      -0.001159  0.027006  0.046259  
Lifespan(years) -0.063539  0.093909 -0.068345  
Wei

In [434]:
resMatrix = {}

df_matrix=pd.DataFrame([])

features = df_edited.columns[df_edited.columns != 'Is_Endangered']
X = df_edited[features]
y = pd.Series(df_edited['Is_Endangered'])

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

ml_classifier = {'LogisticReg' : LogisticRegression(),
             'DecisionTree' : DecisionTreeClassifier(min_samples_split=28),
             'RandomForest' :RandomForestClassifier(),
}

for model in ml_classifier:
    
    clf_model = ml_classifier[model].fit(X_train, y_train)
    y_pred = clf_model.predict(X_test)

    resMatrix['accuracy'] = accuracy_score(y_test, y_pred)
    resMatrix['f1'] = f1_score(y_test, y_pred, average='macro')
    resMatrix['precision'] = precision_score(y_test, y_pred)
    resMatrix['recall'] = recall_score(y_test, y_pred)
    df_matrix=pd.concat([df_matrix,pd.DataFrame(resMatrix,index=[model])],axis=0)
# © Niv Bar for the idea

In [435]:
df_matrix

accuracy        f1  precision    recall
LogisticReg   0.785467  0.606163   0.666667  0.228571
DecisionTree  0.806228  0.694041   0.659091  0.414286
RandomForest  0.826990  0.737082   0.700000  0.500000

# RandomForest Classifier has the best results
with accuracy 0.826990% and f1 0.737082% and precision 0.700000% recall 0.500000%